### PR TITLE
Improve shape inference in the `cat` pipeline 

### PR DESCRIPTION
**Pull request by timholy**
`cat` is frequently called with poor inference, since one only has to concatenate a couple of different container types before inference punts on the result type. While this does not make the return type in mixed container types fully inferrable, it does improve the analysis of the shape.

Because it doesn't fully fix the inference problems, it's not obvious this is a testable improvement except at the aggregate level of backedges. Given that the issues are not fully fixed, the `T::Type` to `::Type{T}` changes don't really seem to be necessary, and I could back those out if desired.

### CODE REVIEW

File: base/abstractarray.jl [hunk 1]
```diff
@@ -1436,15 +1436,15 @@ vcat(V::AbstractVector{T}...) where {T} = typed_vcat(T, V...)
 AbstractVecOrTuple{T} = Union{AbstractVector{<:T}, Tuple{Vararg{T}}}
 
 function _typed_vcat(::Type{T}, V::AbstractVecOrTuple{AbstractVector}) where T
-    n::Int = 0
+    n = 0
     for Vk in V
-        n += length(Vk)
+        n += length(Vk)::Union{Int,UInt}
```
Comment (JeffBezanson) [inline base/abstractarray.jl:1441]: I'm not sure this is safe to assume.

Comment (timholy) [reply base/abstractarray.jl:1441]: I'll change it to `Integer`. That will leave some `promote_rule` invalidations but it maybe be better than nothing.

Comment (timholy) [reply base/abstractarray.jl]: Or do you think we can safely convert-and-assert? Are you wondering about supporting arrays with more elements than `typemax(Int)`?

File: base/abstractarray.jl [hunk 2]
```diff
@@ -1584,17 +1584,18 @@ end
 _cat(dims, X...) = cat_t(promote_eltypeof(X...), X...; dims=dims)
 
 @inline cat_t(::Type{T}, X...; dims) where {T} = _cat_t(dims, T, X...)
-@inline function _cat_t(dims, T::Type, X...)
+@inline function _cat_t(dims, ::Type{T}, X...) where {T}
     catdims = dims2cat(dims)
-    shape = cat_shape(catdims, map(cat_size, X))
+    shape = cat_shape(catdims, map(cat_size, X)::Tuple{Vararg{Union{Int,Dims}}})::Dims
```
Comment (pabloferz) [inline base/abstractarray.jl:1589]: Looking at `cat_size` it seems to me that it would be just `::Union{Int,Dims}`, but maybe I'm missing something.

Comment (timholy) [reply base/abstractarray.jl:1589]: It may not be flattened (yet).

Comment (timholy) [reply base/abstractarray.jl:1589]: Oh, I get your point. Yes, that's essentially what this type-assert says: each `cat_size(X[i])::Union{Int,Dims}`.

Comment (pabloferz) [reply base/abstractarray.jl:1589]: Got it, I somehow missed on the `map` part. LGTM

**Review by KristofferC — APPROVED** [review]

**Pull request merged**